src/components/dropdown/bl-dropdown.ts

```diff
@@ -0,0 +1,170 @@
+import { LitElement, html, CSSResultGroup, TemplateResult } from 'lit';
+import { customElement, property, state, query } from 'lit/decorators.js';
+import {
+  computePosition,
+  flip,
+  offset,
+  autoUpdate,
+  size,
+  MiddlewareArguments,
+} from '@floating-ui/dom';
+import { event, EventDispatcher } from '../../utilities/event';
+import { classMap } from 'lit/directives/class-map.js';
+
+import style from './bl-dropdown.css';
+
+import '../button/bl-button';
+import { ButtonSize, ButtonVariant, ButtonKind } from '../button/bl-button';
+import { ifDefined } from 'lit/directives/if-defined.js';
+
+export type CleanUpFunction = () => void;
+
+export const blDropdownTag = 'bl-dropdown';
+
+/**
+ * @tag bl-dropdown
+ * @summary Baklava Dropdown component
+ */
+@customElement(blDropdownTag)
+export default class BlDropdown extends LitElement {
+  static get styles(): CSSResultGroup {
+    return [style];
+  }
+
+  @query('bl-button')
+  private _dropdownButton: HTMLElement;
+
+  @query('.popover')
+  private _popover: HTMLElement;
+
+  private _cleanUpPopover: CleanUpFunction | null = null;
+
+  @state() private _isPopoverOpen = false;
+
+  /**
+   * Sets the dropdown button label
+   */
+  @property({ type: String, reflect: true })
+  label = 'Dropdown Button';
+
+  /**
+   * Sets the dropdown button variant
+   */
+  @property({ type: String, reflect: true })
+  variant: ButtonVariant = 'primary';
+
+  /**
+   * Sets the dropdown button kind
+   */
+  @property({ type: String, reflect: true })
+  kind: ButtonKind = 'default';
+
+  /**
+   * Sets the dropdown button size
+   */
+  @property({ type: String, reflect: true })
+  size: ButtonSize = 'medium';
+
+  /**
+   * Sets button as disabled
+   */
+  @property({ type: Boolean, reflect: true })
+  disabled = false;
+
+  /**
+   * Fires when dropdown opened
+   */
+  @event('bl-dropdown-open') private onOpen: EventDispatcher<string>;
+
+  /**
+   * Fires when dropdown closed
+   */
+  @event('bl-dropdown-close') private onClose: EventDispatcher<string>;
+
+  connectedCallback() {
+    super.connectedCallback();
+  }
+  disconnectedCallback() {
+    super.disconnectedCallback();
+
+    this._cleanUpPopover && this._cleanUpPopover();
```

Comment:
Do `document.removeEventListener('click', this._handleClickOutside);` if dropdown is open here.

Reply:
I don't think we need this. We already have this when user calls `open()`.

Reply:
If a dropdown is removed from DOM while it's "open", click event listener will stay in memory. I suggested this to prevent that situation.

Reply:
Is there a way to produce this? TBH I don't understand what removing means.

Reply:
In web components `connected` means a copy of this tag is just added to document. And [`disconnected`](https://lit.dev/docs/components/lifecycle/#disconnectedcallback) means it’s removed from DOM. Let’s consider you are using this component in VueJS:

```html
<bl-dropdown v-if=“isAdmin”> …. </bl-dropdown>
```

It’s possible that `isAdmin` can change regarding to another action in the page. And if it switches to `false` when user opened that dropdown, then it’ll be removed from DOM. Currently when you open a dropdown you start listening `click` event of the document. And normally you remove this listener when you close the dropdown but in that case, close couldn’t be happened since it’s completely removed from document. Then that event listener stays in the memory.